### PR TITLE
IPFS Principles

### PR DESCRIPTION
**Pull request by darobin**
This is the PR for the [IPFS Principles](https://hackmd.io/tkHHmdTfTZatyY-35PGVhA?both) draft we have been working on. I'm keeping it as a draft pending further feedback and also because it needs the site to land before it can land.

I hesitated but I didn't file an IPIP with this on the grounds that it doesn't add or modify any implementation. That being said these considerations are potentially load-bearing and this document might end up being referred to a fair bit, so I'm happy to add an IPIP if others prefer it that way!

### CODE REVIEW

File: src/architecture/principles.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+This is intended as a strict requirement in which implements take trust seriously. The point is
+that it's okay to not constantly spend cycles verifying hashes in an internal setup which you
+have reasons to believe is trustworthy. This is *not* a licence to trust an arbitrary data
+source just because you like them.
+
+For instance:
+
+- A JS code snippet that fetches data from an IPFS HTTP gateway without verifying it is not an
+  IPFS implementation.
+- An IPFS HTTP gateway that verifies the data that it is pulling from arbitrary IPFS nodes
```
Comment (autonome) [inline src/architecture/principles.md:170]: `arbitrary` isn't necessary here. maybe `from an IPFS implementation` or `agent`?

File: src/architecture/principles.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+  before serving it over HTTP is an IPFS implementation.
+- That JS piece of code in the first bullet can be turned into an IPFS implementation if it
+  fetches from a :cite[trustless-gateway] and verifies what it gets.
+
+## Self-Certifying Addressability
+
+:dfn[Authority] is control over a given domain of competence and :dfn[naming authority] is
+control over what resources are called.
```
Comment (autonome) [inline src/architecture/principles.md:178]: split into two sentences

File: src/architecture/principles.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+:dfn[Self-certifying addressability] is the property of a naming system such that it is both
+addressable and verifiable: any name is sufficient to interact with a resource, and its mapping
+to that resource can be certified without recourse to additional authority. Self-certifying
+addressability is a key component of a
+[self-certifying web](https://jaygraber.medium.com/web3-is-self-certifying-9dad77fd8d81)
+and it supports capture-resistance which can help mitigate against centralization.
+
+CIDs support :ref[self-certifying addressability]. With CIDs, the authority to name a resource
+resides only with that resource and derives directly from that resource's most intrinsic
```
Comment (autonome) [inline src/architecture/principles.md:195]: variable intrinsicness is edgy but probably not needed here, s/most//

File: src/architecture/principles.md [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+property: its content. This frees interactions with CID-named resources from the power
+relation implicit in a client-server architecture. CIDs are the trust model of IPFS.
+
+An implementation may retrieve a CID without verifying that the resource matches it, but that
+loses the resource's naming authority. Such an implementation would be comparable to an HTTP
+client looking DNS records up from a random person's resolver: it cannot guarantee that the
+addressing is authoritative. Implementers may make informed decisions as to where in their
+systems they support verification, but they should ensure that CIDs are verified whenever
+they also have access to the resource that the CID maps to.
```
Comment (autonome) [inline src/architecture/principles.md:204]: "that CIDs are verified"

Is it the other way round? Eg, making sure content is verified against the CID when you have it?

Comment (autonome) [reply src/architecture/principles.md:204]: or both, depending on execution context?

File: src/architecture/principles.md [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+We tend not to think about addressability because it is so foundational that we
+struggle to apprehend a system without it, but that is precisely why it is important
+that we get it right. You can find extensive historical evidence that TimBL and others saw
+URLs as arguably the most fundamental invention of the Web, and the early groups that
+worked on Web architecture discussed and debated the properties of URLs at length. The
+problems of centralization we face today trace their lineage back to those decisions.
+
+The hierarchical nature of the HTTP addresses was intentional, as TimBL wrote clearly in
+[Web Architecture from 50,000 feet](https://www.w3.org/DesignIssues/Architecture.html):
+> The HTTP space consists of two parts, one hierarchically delegated, for which the
```
Comment (autonome) [inline src/architecture/principles.md:217]: aha, here's `space`. maybe reference this explicitly above in the `IPFS space` comment, in addition to the fixes requested there.

**Review by autonome — CHANGES_REQUESTED** [review]

File: src/architecture/principles.md [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+self-certifying addressability component of the web.
+
+Addressing data using [CIDs](https://github.com/multiformats/cid) is the first defining
+characteristic of IPFS. And the second characteristic, transport-agnosticity, can be
+supported thanks to the verifiability that CIDs offer. Across a vast diversity of implementations,
+architectures, and services, *IPFS is the space of resources that can be interacted with
+over arbitrary transports using a CID*. As Juan Benet once put it,
+"[*That's it!*](https://github.com/multiformats/cid/commit/ece08b40a6b1e9eeafc224e2757d8d1ef3317163#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R43)"
+
+Conversely, any system that exposes interactions with resources based on CIDs is
```
[the remark this answers is not in the document]
Comment (darobin) [reply src/architecture/principles.md:63]: Actually I made it IPFS implementation, it makes more sense here.

File: src/architecture/principles.md [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+environments, and in how intelligence can be applied at the endpoints in novel ways, while
+being strict with outcomes guarantees that the result will be correct and interoperable.
+
+Note that this approach to robustness also covers the
+[End-to-end Principle](https://en.wikipedia.org/wiki/End-to-end_principle).
+
+## IPFS Implementation Requirements
+
+An :dfn[IPFS Implementation]:
+* MUST support addressability using CIDs.
```
[the remark this answers is not in the document]
Comment (darobin) [reply src/architecture/principles.md:120]: "An implementation must implement" feels a bit off?

File: src/architecture/principles.md [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+* SHOULD name all the important resources it exposes using CIDs. Determining which resources are
+  important is a matter of judgment, but anything that another agent might legitimately wish to
+  access is in scope, and it is best to err on the side of inclusion.
+* SHOULD expose the logical units of data that structure a resource (eg. a CBOR document, a file or
+  directory, a branch of a B-tree search index) using CIDs.
+* SHOULD support incremental verifiability, for practical reasons.
+* MAY rely on any transport layer. The transport layer cannot dictate or constrain what IPFS is.
+
+The :dfn[IPFS space] is the set of all CIDs and transport-agnostic operations that can be carried
+out on CIDs.
```
[the remark this answers is not in the document]
Comment (darobin) [reply src/architecture/principles.md:137]: Yeah, it's vestigial, killing it.

File: src/architecture/principles.md [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+
+The simplicity with which one can expose these existing system over IPFS by simply prefixing
+existing addresses to mint CIDs enables radical interoperability with other content-addressable
+systems.
+
+### Verification Matters
+
+The requirements above state that an implementation may forgo verification when "*it is
+possible to ascertain that verification has happened elsewhere in a trusted part of the system.*"
+This is intended as a strict requirement in which implements take trust seriously. The point is
```
[the remark this answers is not in the document]
Comment (darobin) [reply src/architecture/principles.md:161]: I'm still sad that I couldn't work `trustless but verify` in.

File: src/architecture/principles.md (hunk 5 above)
Comment (darobin) [reply src/architecture/principles.md:217]: I think it's extraneous. This is provided for historical context, but it's not like "HTTP space" is used that much. We don't need to copy things that Tim said forever ago :)

File: src/architecture/principles.md [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+use cases and is adaptable to a broad array of operating conditions. Considering cases
+at the boundary may help develop an intuition for the limits that these principles draw.
+
+### Other Content-Addressing Systems
+
+CIDs are readily made compatible with other content-addressable systems, but this does not
+entail that all content-addressable systems are part of IPFS. Git's SHA1 hashes aren't CIDs
+but can be converted into CIDs by prefixing them with `f01781114`. Likewise, BitTorrent v2
+uses multihashes in the `btmh:` scheme. BitTorrent addresses aren't CIDs, but can be
+converted to CIDs by replacing `btmh:` with `f017c`.
```
Comment (aschmahmann) [inline src/architecture/principles.md:151]:
```suggestion
converted to CIDs by replacing `btmh:` with `f017b`.
```

I know I'm the one who told you `c` but it was my mistake 😅. https://blog.libtorrent.org/2020/09/bittorrent-v2/ says `btmh` is for the info-hash and https://github.com/multiformats/multicodec/blob/6f018eb19666553c12955c24b8b5dc29fde6e9f4/table.csv#L48 is the one for info hashes (as described in https://github.com/ipld/js-ipld-torrent-file/pull/3#issuecomment-662137808, which granted is not a great source but unfortunately the code table doesn't always have obvious specs per code listed)

**Comment by aschmahmann** [comment]
Left a small correction, but looks pretty good 😄. Thanks @darobin and everyone who helped make this happen 🙏.

File: src/architecture/principles.md (hunk 10 above)
Comment (darobin) [reply src/architecture/principles.md:151]: LOL, I should have checked — good catch, thanks!

File: src/architecture/principles.md [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+Group (TAG) issue on the topic stated that it "breaks the web".
+
+Federated models only have weak capture-resistance because the federated entities can always
+concede power (precisely because they have ownership) but lack established means to
+support collective organization. As a result, any power imbalance will likely become hard
+to dislodge. A good example is search: as a publisher (the owner of delegated authority
+over your domain) you can cede the rights to index your content but you can't have a voice
+in what is done with the indexed content (individual opt out is not an option). This was
+fine when you could barter content for links, but once search power consolidated, the
+terms of trade deteriorated with no immediate recourse.
```
Comment (lidel) [inline src/architecture/principles.md:275]: :thought_balloon: interestingly, history does not repeat, but rhymes :) 
 [Federated IPNIs](https://github.com/ipni) will fight the same losing battle, and will have to concede power to support arbitrary transports (https://github.com/ipfs/specs/issues/377).

**Review by lidel — APPROVED** [review]
Excellent gift to the IPFS community, thank you @darobin :heart: 

- fixed a typo and one broken link due to https://github.com/ipfs/specs/pull/387
- filled some css papercuts (not blockers, we can fix them in separate PRs: https://github.com/ipfs/spec-generator/issues/8, https://github.com/ipfs/spec-generator/issues/7)
- dropped some :bulb: for posterity ;)

**Review by autonome — COMMENTED** [review]
r+

(x1000)

**Review by autonome — COMMENTED** [review]
maybe this will close it out.

no. apparently you can't close a review on an already merged pr.

my review will live forever unresolved, like a obelisk to the living document it was in service of.

**Pull request merged**